### DISCUSSION TITLE
memory usage

### DISCUSSION THREAD

**Discussion by virtualluke** [Marco-Sulla/python-frozendict#54] [General]
Are there any benchmarks with the difference in memory usage over a standard dict?  I would hope a rather large improvement with the immutable nature.

thanks,
Luke

**Comment by Marco-Sulla** [comment]
Because `tuple` is more compact than `list`? It's not the same case, unluckily. 

frozendict is a map as dict, so it must have anyway a sparse hashtable, or the access by key will be no more O(1). dict was already made more compact in Python 3.6, using the Pypy model.

In theory, one could create a PyFrozendictKeysObject, without:

* dk_refcnt, since it's used by shared dicts, and frozendicts are all compact
* dk_usable, since frozendicts have no hole created by deletion

Furthermore, one could remove from PyFrozendictObject:
* ma_version_tag, since it's immutable
* ma_values, since it's always combined

I don't think it will be a great memory save. These attributes are per dict, not per item, so the memory save will be very low. On the contrary, there's the possibility to make it completely incompatible with dictobject.c functions.

Actually, frozendict is a little more big than dict, since it also has a cache for the hash.
